e2ee: audio support

fippo opened the pull request:
see individual commits, first one adds a checkbox to switch between proper output and middlebox garbage.
The second commit updates the no-encryption scheme to support audio (where chunk.type is undefined; I would prefer to have a chunk.kind but this works) and enables audio in getUserMedia

fippo committed:
e2ee: add checkbox to switch audio output to middlebox and back

in order to avoid playing both proper audio and garbage at the same time

fippo committed:
e2ee: add audio support

updating the no-encryption scheme to use less bytes for VP8 and only
a single byte for audio

fippo commented:
one of the interesting things here is that this shows the weakness of the current (intentionally stupid) crypto scheme -- you can still kinda hear the bip-bop from the fake device even with encryption for short keys.

alvestrand commented on src/content/peerconnection/endtoend-encryption/js/main.js:
```diff
@@ -110,6 +112,19 @@ function dump(chunk, direction, max = 16) {
+//   https://tools.ietf.org/html/rfc6386#section-9.1
+// which is 10 bytes for key frames and 3 bytes for delta frames.
+// For opus (where chunk.type is not set) this is the TOC byte from
+//   https://tools.ietf.org/html/rfc6716#section-3.1
```
Add a description of the purpose, something like:

This allows the middlebox to decode some packets without crashing, making for more interesting pictures and sounds.

fippo commented on src/content/peerconnection/endtoend-encryption/js/main.js:
```diff
@@ -110,6 +112,23 @@ function dump(chunk, direction, max = 16) {
+const frameTypeToCryptoOffset = {
+  key: 10,
+  delta: 3,
+  undefined: 1,
```
I would very much prefer being able to determine audio/video in a way that differs from "does the frame have a type"

alvestrand replied:
You can determine it when installing the function - at that time you know the track type.

fippo commented:
there is a fun caveat here. If you set track.enabled to false and don't have a crypto key, the opus silence frames will be three bytes. Which leads to an out-of-bounds trying to read the checksum. Will fix...

fippo committed:
e2ee: dont throw on opus silence frames

when cryptokey is not set, the sender can generate three byte opus frames for silence.
Don't throw in that case

fippo commented:
Additionally the byteLength can be 0 with DTX when there is silence. Handled correctly :-)
